minor-modes.lisp

```diff
@@ -0,0 +1,1083 @@
+(add-hook *focus-group-hook* 'minor-mode-sync-keys-hook-function)
+
+
+;;;;;;;;;;;;;;;;;;
+;;; Conditions ;;;
+;;;;;;;;;;;;;;;;;;
+
+(define-condition minor-mode-error (error) ())
+
+(define-condition minor-mode-enable-error (minor-mode-error)
+  ((mode :initarg :mode :reader minor-mode-enable-error-mode)
+   (object :initarg :object :reader minor-mode-enable-error-object)
+   (reason :initarg :reason :reader minor-mode-enable-error-reason))
+  (:report
+   (lambda (c s)
+     (format s "Unable to enable minor mode ~A in object ~A.~%Reason: ~A"
+             (minor-mode-enable-error-mode c)
+             (minor-mode-enable-error-object c)
+             (minor-mode-enable-error-reason c)))))
+
+(define-condition minor-mode-disable-error (minor-mode-error)
+  ((mode :initarg :mode :reader minor-mode-disable-error-mode)
+   (object :initarg :object :reader minor-mode-disable-error-object)
+   (reason :initarg :reason :reader minor-mode-disable-error-reason))
+  (:report
+   (lambda (c s)
+     (format s "Unable to disable minor mode ~A in object ~A.~%Reason: ~A"
+             (minor-mode-disable-error-mode c)
+             (minor-mode-disable-error-object c)
+             (minor-mode-disable-error-reason c)))))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Minor Mode Protocol ;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defgeneric minor-mode-global-p (minor-mode-symbol)
+  (:documentation "Return T when MINOR-MODE-SYMBOL denotes a global minor mode")
+  (:method (mode) (declare (ignore mode)) nil))
+
+(defgeneric minor-mode-scope (minor-mode-symbol)
+  (:documentation "Return as a keyword the scope of the minor mode"))
+
+(defgeneric minor-mode-keymap (minor-mode)
+  (:method (minor-mode) nil)
+  (:documentation "Return the top map for the minor mode"))
+
+(defgeneric minor-mode-lighter (mode)
+  (:method (minor-mode) nil)
+  (:method :around (mode)
+    (format nil "~{~A~^ ~}"
+            (call-next-method)))
+  (:documentation "Return a string of minor mode lighters."))
+
+(defgeneric minor-mode-enable-hook (minor-mode-symbol)
+  (:documentation
+   "Returns the minor mode enable hook for a given minor mode symbol. This hook is
+run whenever the minor mode is enabled via autoenable."))
+
+(defmethod no-applicable-method ((f (eql #'minor-mode-enable-hook)) &rest rest)
+  (declare (ignore f rest))
+  nil)
+
+(defgeneric minor-mode-disable-hook (minor-mode-symbol)
+  (:documentation
+   "Returns the minor mode disable hook for a given minor mode symbol.  This hook
+is run whenever the minor mode is disabled via autodisable."))
+
+(defmethod no-applicable-method ((f (eql #'minor-mode-disable-hook)) &rest rest)
+  (declare (ignore f rest))
+  nil)
+
+(defgeneric minor-mode-hook (minor-mode-symbol)
+  (:documentation
+   "Returns the minor mode hook for a given minor mode symbol. This hook is run
+whenever the minor mode is explicitly enabled."))
+
+(defmethod no-applicable-method ((f (eql #'minor-mode-hook)) &rest rest)
+  (declare (ignore f rest))
+  nil)
```

Comment:
The goal for many of these was to default to doing nothing. For this one specifically we want to run the hook returned by this function. But if a user calls this function with a non-minor mode symbol then nil will be returned and interpreted as a hook with no functions. Perhaps it would be better to error if a non minor mode symbol is passed to these hook functions?

Reply:
After thinking about this a little more, the proper course of action is likely to remove these methods and add appropriate restart bind/case and handler bind/case forms to the enabling/disabling/hook-running code. I will work out where that is and after adding that remove these methods.